### PR TITLE
refactor: Don't restrict error type for `Result<T, E>`

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

`Error` can not be right decision for representing error results: for example, we not necessarily want to have stacktraces.

There is no reason why error can't be of any type, including something lightweight like just `true` literal.

This PR refactors error type and also provides some improvements for the `Result` API.

Names has been aligned to `zod` schema library, as it solves very similar problems for us.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Could we not wrap anything in our own Error type?

**Comment by zharinov** [comment]
> Could we not wrap anything in our own Error type?

Please, elaborate on that, I'm not sure I understand what you mean

**Comment by rarkins** [comment]
What's wrong with using Error for everything if we can extend it Birka we like?

**Comment by zharinov** [comment]
Here is the Rubygems-related code I'm refactoring at the moment:

```ts
export type VersionsResult = Result<
  string[],
  'unsupported-api' | 'package-not-found'
>;
```

I think for cases like this, `Error` class is too much:
- we have to declare couple classes separately
- in cases when `Result` is used just for control flow, we don't want stacktraces to be constructed, especially inside loops.

But when we need it, we still can have it I guess.

File: lib/modules/datasource/index.ts [hunk 1]
```diff
@@ -506,12 +505,6 @@ export async function getPkgReleases(
   return res;
 }
 
-export function getPkgReleasesSafe(
```
Comment (viceice) [inline lib/modules/datasource/index.ts:509]: looks like like unrelated change?

Comment (zharinov) [reply lib/modules/datasource/index.ts:509]: It's related to the place when this function was called. When I simplified it's calling code, turned out this function isn't needed anymore. And I needed to change calling code at the `lookup` function anyway. I think nothing will break after "inlining" of this function.

**Comment by zharinov** [comment]
Actually, maybe the single hierarchy of errors is the good thing, I just don't want them to be descendants of `Error` class. Hierarchy of nested union types (`Error` class included) looks good to me, though.

File: lib/modules/datasource/index.ts (hunk 1 above)
Comment (viceice) [reply lib/modules/datasource/index.ts:509]: @rarkins added this function to leave the API compatible for our own tools. this will now break and I need to fix it after merge 😏

Comment (rarkins) [reply lib/modules/datasource/index.ts:509]: Is it ok for you?

Comment (viceice) [reply lib/modules/datasource/index.ts:509]: sure, just was a little confused because the PR title and description told me something different than removing this function.

**Review by viceice — APPROVED** [review]
